plugins/inputs/t128_graphql/query_builder.go

```diff
@@ -1,20 +1,162 @@
 package t128_graphql
 
-import "strings"
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
 
-func buildQuery(entryPoint string, fields map[string]string, tags map[string]string) string {
-	var replacer = strings.NewReplacer("[", "(", "]", "\")", "/", "{", ":", ":\"")
-	query := "query {" + replacer.Replace(entryPoint) + "{"
+	"github.com/Jeffail/gabs"
+)
 
-	for _, element := range fields {
-		query += "\n" + element
-	}
-	query = strings.TrimSpace(query)
-	for _, element := range tags {
-		query += "\n" + element
+const (
+	predicateTag = "$"
+)
+
+/*
+BuildQuery first creates an intermediary query object that is traversed by buildQueryBody() in pre-order
+
+Args:
+	entryPoint example - "allRouters[name:ComboEast]/nodes/nodes[name:combo-east]/nodes/arp/nodes"
+	fields example - map[string]string{"enabled": "enabled"}
+	tags example - map[string]string{
+			"name": "name",
+			"admin-status":  "state/adminStatus",
+		}
+
+Example:
+	For the example input above, buildQueryObject() will produce the following query object
+
+	{
+		"allRouters": {
+			"$name":"ComboEast",
+			"nodes": {
+				"nodes": {
+					"$name":"combo-east",
+					"nodes": {
+						"arp": {
+							"nodes": {
+								"enabled":"enabled",
+								"name":"name",
+								"state": {
+									"adminStatus":
+									"admin-status"
+								}
+							}
+						}
+					}
+				}
+			}
+		}
 	}
 
-	query += "\n" + strings.Repeat("}", strings.Count(query, "{"))
+	And buildQueryBody() will build the following
+
+	{
+	allRouters(name:"ComboEast"){
+	nodes{
+	nodes(name:"combo-east"){
+	nodes{
+	arp{
+	nodes{
+	enabled
+	name
+	state{
+	adminStatus}}}}}}}}
+*/
+func BuildQuery(entryPoint string, fields map[string]string, tags map[string]string) string {
+	query := "query "
 
+	var buf bytes.Buffer
+	w := io.Writer(&buf)
+	jsonObj := buildQueryObject(entryPoint, fields, tags)
+	buildQueryBody(jsonObj, w)
+
+	query += buf.String()
 	return query
 }
+
+//buildQueryBody creates an intermediary query object that is traversed by buildQueryBody
+func buildQueryObject(entryPoint string, fields map[string]string, tags map[string]string) *gabs.Container {
+	jsonObj := gabs.New()
+
+	parsedEntryPoint := ParseEntryPoint(entryPoint)
+
+	addToQueryObj(jsonObj, parsedEntryPoint.Predicates, true, false, "")
+	addToQueryObj(jsonObj, fields, false, true, parsedEntryPoint.QueryPath)
+	addToQueryObj(jsonObj, tags, false, true, parsedEntryPoint.QueryPath)
+
+	return jsonObj
+}
+
+func addToQueryObj(jsonObj *gabs.Container, items map[string]string, keyIsPath bool, usesSlash bool, basePath string) {
+	var replacer = strings.NewReplacer("/", ".")
+	for key, value := range items {
+		if keyIsPath {
+			if usesSlash {
+				partialPath := replacer.Replace(key)
+				jsonObj.SetP(value, basePath+partialPath)
+			} else {
+				jsonObj.SetP(value, basePath+key)
+			}
```

Comment:
Unless I'm mistaken, this `if/else` block here and below are identical, correct? Can we maybe define a local function/lambda (sorry using the C++ lingo 😄 ) for this instead?

Reply:
Looks like go supports lambda functions. I'll give them a shot as a way to shorten this function

Reply:
Could you swap the key and values for the predicates so that they can all do the same thing? Only reason I could see not is if the values aren't unique and thus couldn't be keys.

Something maybe a little cleaner than:
```
func swap(m map[string]string) map[string]string {
     new = make(map[string]string, len(m))
     for k, v in range m {
           new[v] = k
     }
     return new
}
```
Or better yet, when parsing the base path's predicates, could put them in the map this way?

Reply:
Some of the predicate names aren't unique like "name" so I went with the inner function approach to start